core/src/main/java/com/exadel/aem/backpack/core/servlets/UploadPackageServlet.java

```diff
@@ -0,0 +1,128 @@
+
+import javax.jcr.Session;
+import javax.servlet.Servlet;
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import static com.exadel.aem.backpack.core.servlets.BuildPackageServlet.APPLICATION_JSON;
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static javax.servlet.http.HttpServletResponse.SC_CONFLICT;
+
+/**
+ * Serves as the network endpoint for user requests that trigger package creation<br><br>
+ * <p>
+ * See also:<br>
+ * {@link BuildPackageServlet} - endpoint for requests for building of created package and reporting package status<br>
+ * {@link PackageInfoServlet} - endpoint for requests for information on previously built packages
+ */
+@Component(
+        service = Servlet.class,
+        property = {
+                "sling.servlet.paths=/services/backpack/uploadPackage",
+                "sling.servlet.methods=[get,post]"
+        })
+@SuppressWarnings("PackageAccessibility")
+// because Servlet and HttpServletResponse classes reported as a non-bundle dependency
+public class UploadPackageServlet extends SlingAllMethodsServlet {
+
+    private static final String PARAM_FORCE_UPDATE = "forceUpdate";
+    private static final String PARAM_FILEUPLOAD = "fileupload";
+    private static final long serialVersionUID = 1L;
+    private static final Gson GSON = new Gson();
+
+    @Reference
+    @SuppressWarnings("UnusedDeclaration") // value injected by Sling
+    private transient PackageService packageService;
+
+    /**
+     * Processes {@code POST} requests to the current endpoint. Attempts to create a package for a consequential build
+     * according to the request parameters.
+     * Request parameters are parsed to a {@link PackageModel} which is validated and passed
+     * to the corresponding {@link PackageService} routine if proven valid; otherwise, the {@code HTTP status 400} reported
+     *
+     * @param request  {@code SlingHttpServletRequest} instance
+     * @param response {@code SlingHttpServletResponse} instance
+     * @throws IOException in case writing data to the {@code SlingHttpServletResponse} fails
+     */
+    @Override
+    protected void doPost(final SlingHttpServletRequest request,
+                          final SlingHttpServletResponse response) throws IOException {
+        Session session = request.getResourceResolver().adaptTo(Session.class);
+        byte[] fileUploadBytesArray = getFileUploadBytesArray(request);
+        boolean forceUpdate = getForceUpdate(request);
+
+        JcrPackageWrapper jcrPackageWrapper = packageService.uploadPackage(session, fileUploadBytesArray, forceUpdate);
+
+        writeResponse(response, jcrPackageWrapper);
+    }
+
+    private byte[] getFileUploadBytesArray(final SlingHttpServletRequest request) {
+        String parameter = request.getParameter(PARAM_FILEUPLOAD);
+        if (parameter != null) {
+            return parameter.getBytes(ISO_8859_1);
```

Comment:
Are you sure that it will be always ISO_8859_1? Maybe it's better to create a config file?

Reply:
It doesn't work with other encoding(These data were obtained experimentally).